Pipeline Results

abhinandan13jan opened the pull request:
# Adresses
https://issues.redhat.com/browse/ODC-3765

# Screenshots
![Screenshot from 2021-02-17 15-02-19](https://user-images.githubusercontent.com/24852534/108187302-5304e580-7134-11eb-9243-cdc0cab1ceb0.png)
![Screenshot from 2021-02-15 19-34-56](https://user-images.githubusercontent.com/24852534/107956499-40b86980-6fc5-11eb-8448-2dc261930953.png)
![Screenshot from 2021-02-15 19-35-04](https://user-images.githubusercontent.com/24852534/107956504-41510000-6fc5-11eb-80e8-d0d7b2b0cb06.png)


# Data for test

Add the following task and pipeline and then start the Pipeline using the modal and add parameters of your choice. It will return the results.
```
apiVersion: tekton.dev/v1beta1
kind: Task
metadata:
  name: add-task
spec:
  params:
    - name: first
      description: the first operand
    - name: second
      description: the second operand
  results:
    - name: sum
      description: the sum of the first and second operand
  steps:
    - name: add
      image: alpine
      env:
        - name: OP1
          value: $(params.first)
        - name: OP2
          value: $(params.second)
      command: ["/bin/sh", "-c"]
      args:
        - echo -n $((${OP1}+${OP2})) | tee $(results.sum.path);
---
apiVersion: tekton.dev/v1beta1
kind: Pipeline
metadata:
  name: sum-three-pipeline
spec:
  params:
    - name: first
      description: the first operand
    - name: second
      description: the second operand
    - name: third
      description: the third operand
  tasks:
    - name: first-add
      taskRef:
        name: add-task
      params:
        - name: first
          value: $(params.first)
        - name: second
          value: $(params.second)
    - name: second-add
      taskRef:
        name: add-task
      params:
        - name: first
          value: $(tasks.first-add.results.sum)
        - name: second
          value: $(params.third)
  results:
    - name: sum
      description: the sum of all three operands
      value: $(tasks.second-add.results.sum)
    - name: partial-sum
      description: the sum of first two operands
      value: $(tasks.first-add.results.sum)
    - name: all-sum
      description: the sum of everything
      value: $(tasks.second-add.results.sum)-$(tasks.first-add.results.sum)
---
apiVersion: tekton.dev/v1beta1
kind: PipelineRun
metadata:
  name: sum-three-pipeline-run
spec:
  pipelineRef:
    name: sum-three-pipeline
  params:
    - name: first
      value: "2"
    - name: second
      value: "10"
    - name: third
      value: "10"
```
# Tests
Added new test Suite `ResultsList.spec.tsx`

# Browser Conformance
Chrome 88

andrewballantyne commented:
/cc @andrewballantyne

andrewballantyne commented:
/cc @karthikjeeyar

andrewballantyne commented:
/assign

abhinandan13jan commented on frontend/packages/pipelines-plugin/src/components/pipelineruns/detail-page-tabs/PipelineRunDetails.tsx:
```diff
@@ -29,47 +30,57 @@ export const PipelineRunDetails: React.FC<PipelineRunDetailsProps> = ({ obj: pip
       </div>
-    </div>
+      {pipelineRun.status?.pipelineResults && (
+        <div className="co-m-pane__body">
```
The designs(Doc) as well as details pages like `PodDetails` have multiple sections with "co-m-pane__body". Hence I have created this keeping those consistent. Let me know what should be done?

andrewballantyne replied:
Fair enough... I didn't run it, I was just looking at the code.

abhinandan13jan commented on frontend/packages/pipelines-plugin/src/components/pipelineruns/detail-page-tabs/PipelineRunDetails.tsx:
```diff
@@ -29,47 +30,57 @@ export const PipelineRunDetails: React.FC<PipelineRunDetailsProps> = ({ obj: pip
-            namespace={pipelineRun.metadata.namespace}
-          />
-          {pipelineRefExists(pipelineRun) && (
+    <>
```
Created sub components

abhinandan13jan commented on frontend/packages/pipelines-plugin/src/components/taskruns/TaskRunDetails.tsx:
```diff
@@ -17,43 +18,52 @@ export interface TaskRunDetailsProps {
-            namespace={taskRun.metadata?.namespace}
-          />
-          {taskRun?.status?.podName && (
+    <>
```
updated

abhinandan13jan commented on frontend/packages/pipelines-plugin/src/components/shared/results/ResultsList.tsx:
```diff
@@ -0,0 +1,30 @@
+const ResultsList: React.FC<ResultsListProps> = ({ results, resourceKind }) => {
+  const { t } = useTranslation();
+  if (!results.length) return null;
+  const columns = [{ title: t('Name') }, { title: t('Value') }];
```
updated

andrewballantyne commented on frontend/packages/pipelines-plugin/src/types/coreTekton.ts:
```diff
@@ -11,3 +11,14 @@ export type TektonResource = {
+export type TektonResultsDefinition = {
+  name: string;
+  description?: string;
+  value: string;
```
I think value is optional 🤔 

It's like as if you were to craft the response vs just allowing it to be whatever is jammed into it during execution.

abhinandan13jan replied:
Just confirmed only description is optional. value must be there in definition.

andrewballantyne replied:
Your example in your description says otherwise:
```
apiVersion: tekton.dev/v1beta1
kind: Task
metadata:
  name: add-task
spec:
  params:
    - name: first
      description: the first operand
    - name: second
      description: the second operand
  results:
    - name: sum
      description: the sum of the first and second operand
  steps:
    - name: add
      image: alpine
      env:
        - name: OP1
          value: $(params.first)
        - name: OP2
          value: $(params.second)
      command: ["/bin/sh", "-c"]
      args:
        - echo -n $((${OP1}+${OP2})) | tee $(results.sum.path);`
```

andrewballantyne commented on frontend/packages/pipelines-plugin/src/utils/pipeline-utils.ts:
```diff
@@ -457,3 +458,17 @@ export const pipelinesTab = (kindObj: K8sKind) => {
+    rows: results.map((result) => ({ cells: [result.name, result.value] })),
+    columns: ['pipelines-plugin~Name', 'pipelines-plugin~Value'],
+  };
+};
```
Did my idea not work? I thought PF accepts either strings or objects. I'll admit I didn't test it. I think there is value in the [Tuple](https://www.typescriptlang.org/docs/handbook/basic-types.html#tuple) though to help constrain headers and body to the same length of an array... otherwise I bet the Table will do its best and likely render more than 2 columns.

Maybe we keep the render aspects out of this utility like you did for the columns, at least that way we can keep this pure 🤔 

Thoughts?

abhinandan13jan replied:
Keeping the controls of column in the Component makes more sense. Moved it.

abhinandan13jan replied:
updated, yes. it works with the tuples. Did pass my local lint although I got some errors in my IDE

abhinandan13jan commented on frontend/packages/pipelines-plugin/src/components/shared/results/ResultsList.tsx:
```diff
@@ -0,0 +1,30 @@
+  return (
+    <>
+      <SectionHeading text={t('{{resource}} Results', { resourceKind })} />
+      <Table aria-label={t('{{resource}} Results', { resourceKind })} cells={columns} rows={rows}>
```
updated. yes this one should be namespaced

abhinandan13jan commented on frontend/packages/pipelines-plugin/locales/en/pipelines-plugin.json:
```diff
@@ -12,9 +12,9 @@
-  "Pipeline Run details": "Pipeline Run details",
   "Status": "Status",
   "Pipeline": "Pipeline",
+  "Pipeline Run details": "Pipeline Run details",
```
P.S: this was done by i18n command

abhinandan13jan commented:
cc: @bgliwa01 @karthikjeeyar

andrewballantyne commented on frontend/packages/pipelines-plugin/src/types/coreTekton.ts:
```diff
@@ -11,3 +11,14 @@ export type TektonResource = {
   type: string; // TODO: limit to known strings
 };
+
+export type TektonResultsDefinition = {
```
Let's delete this type and add it directly to the `Pipeline` and `Task`. 

- Pipeline needs all three `name`, `description?` and `value`
- Task needs just two `name` and `description?`

abhinandan13jan replied:
updated

andrewballantyne commented on frontend/packages/pipelines-plugin/src/types/task.ts:
```diff
@@ -16,5 +16,6 @@ export type TaskKind = K8sResourceCommon & {
       image?: string;
       resources?: {}[];
     }[];
+    results?: TektonResultsDefinition[];
```
Inline this per my last comment.

andrewballantyne commented on frontend/packages/pipelines-plugin/src/types/pipeline.ts:
```diff
@@ -69,6 +69,7 @@ export type PipelineSpec = {
   serviceAccountName?: string;
   tasks: PipelineTask[];
   workspaces?: PipelineWorkspace[];
+  results?: TektonResultsDefinition[];
```
Inline this per my (second) last comment

andrewballantyne reviewed (changes requested):
Per discussion on slack, these are the requested type changes.

andrewballantyne reviewed (commented):
@abhinandan13jan I've not successfully made your examples run. I get a failure on the first task:
> /bin/sh: syntax error: EOF in backquote substitution

It does make the PipelineRun page not look great when it fails. @bgliwa01 
<img width="876" alt="Screen Shot 2021-02-11 at 8 18 13 AM" src="https://user-images.githubusercontent.com/8126518/107642989-94345b80-6c43-11eb-84c7-f9280bd33656.png">

Is this a desired outcome for a failed PipelineRun's results? Do we just print them verbatim?

andrewballantyne commented:
@abhinandan13jan What were your values you supplied to the Start Pipeline Modal?

andrewballantyne commented on frontend/packages/pipelines-plugin/src/components/pipelineruns/detail-page-tabs/PipelineRunDetails.tsx:
```diff
@@ -1,75 +1,28 @@
+        <div className="co-m-pane__body">
+          <ResultsList
+            results={pipelineRun.status?.pipelineResults}
+            resourceKind={pipelineRun.kind}
```
```suggestion
            resourceKind={pipelineRun.label}
```

abhinandan13jan replied:
I assumed u meant `PipelineRunModel.labe`l and updated accordingly

andrewballantyne replied:
Right you are, my bad! Good catch.

andrewballantyne commented on frontend/packages/pipelines-plugin/locales/en/pipelines-plugin.json:
```diff
@@ -196,6 +196,7 @@
   "Select a Project to view the list of Pipelines": "Select a Project to view the list of Pipelines",
   "Create Pipeline": "Create Pipeline",
   "by name": "by name",
+  "{{resource}} Results": "{{resource}} Results",
```
```suggestion
  "{{resourceName}} Results": "{{resourceName}} Results",
```
Suggest upping the variable name used here as the translators could use it to help them understand context.

andrewballantyne commented on frontend/packages/pipelines-plugin/src/components/taskruns/TaskRunDetails.tsx:
```diff
@@ -1,59 +1,25 @@
-    </div>
+      {taskRun.status?.taskResults && (
+        <div className="co-m-pane__body">
+          <ResultsList results={taskRun.status?.taskResults} resourceKind={taskRun.kind} />
```
```suggestion
          <ResultsList results={taskRun.status?.taskResults} resourceKind={taskRun.label} />
```

andrewballantyne reviewed (changes requested):
I noticed while watching your video that the header above the Visualization has a space between the `Pipeline` and `Run` words... but yours does not for Results... I think we should keep them consistent and use the `label` instead of the `kind` name.

bgliwa01 commented:
Is there a reason there isn't a space in between Pipeline and Run in "PipelineRun Results"? Per the design (which is a high level design), it should be "Pipeline Run results".

bgliwa01 commented:
> @abhinandan13jan I've not successfully made your examples run. I get a failure on the first task:
> 
> > /bin/sh: syntax error: EOF in backquote substitution
> 
> It does make the PipelineRun page not look great when it fails. @bgliwa01
> <img alt="Screen Shot 2021-02-11 at 8 18 13 AM" width="876" src="https://user-images.githubusercontent.com/8126518/107642989-94345b80-6c43-11eb-84c7-f9280bd33656.png">
> 
> Is this a desired outcome for a failed PipelineRun's results? Do we just print them verbatim?

Should we even be showing results of a failed Pipeline Run? To me, it looks like there is not an actual value in the "value" column so I don't know if this information would even be useful/necessary to show @siamaksade @serenamarie125

andrewballantyne commented on frontend/packages/pipelines-plugin/src/utils/pipeline-utils.ts:
```diff
@@ -457,3 +458,17 @@ export const pipelinesTab = (kindObj: K8sKind) => {
+
+export const getCellsFromResults = (results: TektonResultsRun[]): ResultCells => {
+  return {
+    columns: ['pipelines-plugin~Name', 'pipelines-plugin~Value'],
```
```
    // t('pipelines-plugin~Name') t('pipelines-plugin~Value')
```
This can help us avoid having to bring `t` into the utility and still maintain the fields in the json translation file.

abhinandan13jan replied:
Thanks. Updated accordingly

abhinandan13jan commented:
> Is there a reason there isn't a space in between Pipeline and Run in "PipelineRun Results"? Per the design (which is a high level design), it should be "Pipeline Run results". Same for Task Run results.

Hi @bgliwa01 I've fixed the spacing between Pipeline and Run by using the label now instead of Kind. I've updated the screenshots
@andrewballantyne Updated the test data with PipelineRun data, should work now. 

Let me know whether we need to hide `Results` when PipelineRun / TaskRun fails?

![Screenshot from 2021-02-12 14-51-44](https://user-images.githubusercontent.com/24852534/107750786-81ed1700-6d42-11eb-9305-4bb73f986c5b.png)
![Screenshot from 2021-02-12 14-51-53](https://user-images.githubusercontent.com/24852534/107750792-831e4400-6d42-11eb-973a-93571b7caecb.png)

bgliwa01 commented:
Results should be lowercase so it would be "Pipeline Run results" and "Task Run results"

andrewballantyne commented:
Good catch, I did not see that @bgliwa01

abhinandan13jan commented:
> Results should be lowercase so it would be "Pipeline Run results" and "Task Run results"
Updated `results` to lower case also added an emptystate
![Screenshot from 2021-02-15 18-26-22](https://user-images.githubusercontent.com/24852534/107956347-0cdd4400-6fc5-11eb-94a2-fab76fc226f3.png)
cc: @andrewballantyne @bgliwa01

andrewballantyne commented on frontend/packages/pipelines-plugin/src/components/shared/results/ResultsList.tsx:
```diff
@@ -0,0 +1,34 @@
+  return (
+    <>
+      <SectionHeading text={t('pipelines-plugin~{{resourceName}} Results', { resourceName })} />
+      <Table
```
Did you push code to make the alert appear? I don't see it 😕

abhinandan13jan replied:
My bad.. updated now

andrewballantyne commented on frontend/packages/pipelines-plugin/locales/en/pipelines-plugin.json:
```diff
@@ -196,6 +196,8 @@
   "Select a Project to view the list of Pipelines": "Select a Project to view the list of Pipelines",
   "Create Pipeline": "Create Pipeline",
   "by name": "by name",
+  "{{resource}} results": "{{resource}} results",
+  "No {{resource}} results available due to failure": "No {{resource}} results available due to failure",
```
Hmmm... thought we agreed on `resourceName` to help with the translators...

abhinandan13jan replied:
Ok, I'll make it `resourceName` for all the related ones.

andrewballantyne replied:
You changed the usages but not the definition... you still need to update the json file.

andrewballantyne reviewed (changes requested):
I'll fire this up and see if this is good to merge. I think we should go with `resourceName` instead of `resource` though... I may be misremembering, but I thought you did this in an earlier change 🤔

andrewballantyne reviewed (approved):
Looks good, we'll need to probably swing back with changes for message.

/assign @karthikjeeyar

karthikjeeyar commented on frontend/packages/pipelines-plugin/src/components/shared/results/ResultsList.tsx:
```diff
@@ -0,0 +1,47 @@
+import * as React from 'react';
+import { Alert } from '@patternfly/react-core';
+import { useTranslation } from 'react-i18next';
```
nit: sort imports

abhinandan13jan replied:
updated

karthikjeeyar commented on frontend/packages/pipelines-plugin/src/components/shared/results/ResultsList.tsx:
```diff
@@ -0,0 +1,47 @@
+          <TableBody />
+        </Table>
+      ) : (
+        <Alert
```
It would be nice, if we can include a test for this component where it conditionally renders the `Table` vs `Alert` based on the `runStatus`

abhinandan13jan replied:
added unit tests

karthikjeeyar commented on frontend/packages/pipelines-plugin/src/utils/__tests__/pipeline-test-data.ts:
```diff
@@ -290,3 +290,59 @@ export const mockRunDurationTest = [
   pipelineTestData[PipelineExampleNames.CLUSTER_PIPELINE].pipelineRuns[DataState.SUCCESS],
 ];
+
+export const taskRunWithResults: TaskRunKind = {
+  apiVersion: 'tekton.dev/v1beta1',
+  kind: 'TaskRun',
```
Please move this test data to `taskrun-test-data.ts` (This file contains all the taskrun test data).

abhinandan13jan replied:
moved

jerolimov reviewed (approved):
Tested this locally and it works fine. Nice to see this new tests.

/lgtm
